src/chips/nrf51/src/clock.rs

```diff
@@ -0,0 +1,180 @@
+//! The clock peripheral of the nRF51 series (chapter 13 of
+//! the nRF51 reference manual v3.0), which manages the 
+//! low frequency and high frequency clocks. The low frequency
+//! clock drives the real time clock (RTC), while the 
+//! high frequency clocks drive the timer system.
+//!
+//! Author: Philip Levis
+//! Date: August 18, 2016
+
+use common::VolatileCell;
+use common::take_cell::TakeCell;
+use core::mem;
+
+struct Registers {
+    pub tasks_hfclkstart    : VolatileCell<u32>,
+    pub tasks_hfclkstop     : VolatileCell<u32>,
+    pub tasks_lfclkstart    : VolatileCell<u32>,
+    pub tasks_lfclkstop     : VolatileCell<u32>,
+    pub tasks_cal           : VolatileCell<u32>,
+    pub tasks_cstart        : VolatileCell<u32>,
+    pub tasks_cstop         : VolatileCell<u32>,
+    _reserved1: [VolatileCell<u32>; 57],
+    pub events_hfclkstarted : VolatileCell<u32>,
+    pub events_lfclkstarted : VolatileCell<u32>,
+    pub done                : VolatileCell<u32>,
+    pub ctto                : VolatileCell<u32>,
+    _reserved2: [VolatileCell<u32>; 125],
+    pub intenset            : VolatileCell<u32>,
+    pub intenclr            : VolatileCell<u32>,
+    _reserved3: [VolatileCell<u32>; 63],
+    pub hfclkrun            : VolatileCell<u32>,
+    pub hfclkstat           : VolatileCell<u32>,
+    _reserved4: [VolatileCell<u32>; 1],
+    pub lfclkrun            : VolatileCell<u32>,
+    pub lfclkstat           : VolatileCell<u32>,
+    pub lfclksrccopy        : VolatileCell<u32>,
+    _reserved5: [VolatileCell<u32>; 62],
+    pub lfclksrc            : VolatileCell<u32>,
+    _reserved6: [VolatileCell<u32>; 7],
+    pub ctiv                : VolatileCell<u32>,
+    _reserved7: [VolatileCell<u32>; 5],
+    pub xtalfreq            : VolatileCell<u32>,
+}
+
+const CLOCK_BASE: usize = 0x40000000;
+
+pub enum InterruptField {
+    HFCLKSTARTED = (1 << 0),
+    LFCLKSTARTED = (1 << 1),
+    DONE         = (1 << 3),
+    CTTO         = (1 << 4),
+}
+
+pub enum ClockTaskTriggered {
+    NO   = 0,
+    YES  = 1,
+}
+
+pub enum ClockRunning {
+    NORUN   = 0,
+    RUN     = (1 << 16),
+}
+
+pub enum LowClockSource {
+    RC            = 0,
+    XTAL          = 1,
+    SYNTH         = 2,
+    MASK          = 0x3,
+}
+
+pub enum HighClockSource {
+    RC            = 0,
+    XTAL          = 1,
+}
+
+pub enum XtalFreq {
+    F16MHz         = 0xFF,
+    F32MHz         = 0,
+}
+
+pub struct Clock {
+    client: TakeCell<&'static ClockClient>,
+}
+
+pub trait ClockClient {
+    /// All clock interrupts are control signals, e.g., when
+    /// a clock has started etc. We don't actually handle any
+    /// of them for now, but keep this trait in place for if we
+    /// do need to in the future.
+    fn event(&self);
+}
+
+pub static mut CLOCK : Clock = Clock {
+    client: TakeCell::empty(),
+};
+
+
+impl Clock {
+    #[allow(non_snake_case)]
+    fn CLOCK(&self) -> &'static Registers {
```

Comment:
WRT StaticRef:

We haven't converged on a way of properly lifting pointers to register structures in Rust references. The purpose of `StaticRef` is to figure out how to require significantly less code to be trusted, where so far the methods we've been using haven't covered 100% of cases:

  * Throughout the SAM4L codebase we use `*const Registers` fields in the device struct and cast explicitly in each method. This works across devices with a single register base (e.g. the AST) as well as those with multiple (e.g. USART0, USART1, etc). However, it requires an unsafe cast in each method.

  * In the NRF51 @lizardo and you seem to be using the `DEVICE_NAME()` style, which isolates the `unsafe` code, but won't work as elegantly for device types with multiple instances (not sure if there are any on the nrf51).

  * StaticRef is the (almost) best of both worlds (it technically requires an extra word of memory for singleton devices relative to the second option above), but I haven't tested a full device with it yet.

In these cases, I think it's fine to try an isolated example. If it doesn't work out well, it's trivial to port it to the other mechanisms, and I will. If it does, and we decide to adopt it, I'll go through and change everything at once (also trivial, just touches a lot of files).